### PR TITLE
homework

### PR DESCRIPTION
**Pull request by Stasgavriluk**

### CODE REVIEW

File: cypress/e2e/bank_account.spec.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    })
+
+    // successful onboarding
+    it('should successful onboarding', () => {
+        cy.ui_onboarding()
+    })
+
+    // delete bank account
+    it('should delete bank account', () => {
+        cy.ui_logout()
```
Comment (ievgendco) [inline cypress/e2e/bank_account.spec.js]: Your tests are dependent!!! You should not rely on the result of previous tests. ALL TESTS SHOULD BE INDEPENDENT

File: cypress/e2e/sign_in_page.spec.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        cy.reload()
+        // cy.get('.makeStyles-root-1').click()   if we click anywhere button will be disabled
+        cy.get(sign_in_page.signin_submit).should('be.disabled') //button is not disabled by default
+    })
+
+    it('should have Don\'t have an account? Sign Up clickable link under Sign in btn', () => {
+        cy.get(sign_in_page.link).should('be.visible').and('have.attr', 'href', '/signup')
+            .click()
+        cy.url().should('contain', '/signup')
+        cy.go("back")
```
Comment (ievgendco) [inline cypress/e2e/sign_in_page.spec.js]: Do you really need this step in your test?

File: cypress/e2e/signup_signin_logout.spec.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        cy.url().should('contain', '/signup')
+        })
+
+    // errors for sign-up page
+    it('should show validation errors for first name field', () => {
+        cy.get(sign_up_page.first_name_validation_message).should('not.exist')
+        cy.get(sign_up_page.first_name_field).click().blur()
+        cy.get(sign_up_page.first_name_validation_message).should('be.visible').and('have.text', 'First Name is required')
+    })
+    it('should show validation errors for last name field', () => {
```
Comment (ievgendco) [inline cypress/e2e/signup_signin_logout.spec.js:22]: Common practice: One single space between every single test!

File: cypress/e2e/signup_signin_logout.spec.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    // login
+    it('should allow a visitor to login', () => {
+        cy.ui_login(user_info.username, user_info.password)
+    })
+
+    //logout
+    it('should allow a visitor to logout', () => {
+        cy.ui_logout()
+    })
+
```
Comment (ievgendco) [inline cypress/e2e/signup_signin_logout.spec.js:69]: Please remove unnecessary spaces

File: cypress/e2e/signup_signin_logout.spec.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        cy.reload()
+        cy.get(sign_in_page.username_field).type('bobross')
+        cy.get(sign_in_page.password_field).type('566').blur()
+        cy.get(sign_in_page.password_validation_message).should('be.visible').and('have.text', 'Password must contain at least 4 characters')
+        cy.get(sign_in_page.password_field).type('5465').blur()
+        cy.get(sign_in_page.password_validation_message).should('not.exist')
+    })
+    })
+
+// Homework 21.07
```
Comment (ievgendco) [inline cypress/e2e/signup_signin_logout.spec.js]: Do you or your team need these comments?

File: cypress/support/commands.js [hunk 6]
```diff
@@ -1,25 +1,50 @@
...
-// Cypress.Commands.overwrite('visit', (originalFn, url, options) => { ... })
\ No newline at end of file
+import {sign_in_page, user_info} from "../selectors/sign_in_page";
+import {main_page} from "../selectors/main_page.selector";
+
+Cypress.Commands.add("ui_login", (username, password, { rememberUser = false } = {}) => {
+    const signinPath = "/signin"
+    cy.intercept("POST", "/login").as("loginUser");
+    cy.location("pathname", { log: false }).then((currentPath) => {
+        if (currentPath !== signinPath) {
```
Comment (ievgendco) [inline cypress/support/commands.js]: Please read about single line statements

**Review by ievgendco — CHANGES_REQUESTED** [review]

File: cypress/e2e/bank_account.spec.js (hunk 1 above)
Comment (Stasgavriluk) [reply cypress/e2e/bank_account.spec.js]: It didnt create bank account after create it. We should logout and then login into account to see the bank accounts page with our bank account.

File: cypress/e2e/new-transaction.spec.js [hunk 7]
```diff
@@ -1,97 +1,156 @@
...
+        cy.ui_logout()
+        cy.ui_login(receiverUserName, password)
+        cy.url().should("not.contain", "/signin")
+        cy.get(transaction.user_balance).should(($el) => {
+            expect($el.text()).to.not.equal(receiverStartBalance)
+        })
     })
 
-    it("searches for a user by attribute", function () {
+    it("submits a transaction request and accepts the request for the receiver", () => {
```
Comment (anatoly-devico) [inline cypress/e2e/new-transaction.spec.js:127]: This test doesn't work

File: cypress/selectors/sign_in_page.js [hunk 8]
```diff
@@ -29,17 +29,3 @@ export const sign_in_page = {
     log_out: "[data-test='sidenav-signout']",
     sidenav_username: "[data-test='sidenav-username']"
 }
-
```
Comment (anatoly-devico) [inline cypress/selectors/sign_in_page.js]: Rename "should show disabled by default sign in btn" test to "should show enabled by default sign in btn" and refactor it due to new name.

File: cypress/e2e/signup_signin_logout.spec.js [hunk 9]
```diff
@@ -48,12 +51,12 @@ describe('UI test for sign-up, login and logout', () => {
 
     // sign-up
     it('should allow a visitor to sign-up', () => {
-        cy.ui_sign_up()
+        cy.ui_sign_up(userName, password)
```
Comment (anatoly-devico) [inline cypress/e2e/signup_signin_logout.spec.js:54]: In this spec following tests don't work:

1. should allow a visitor to logout
2. should show login errors for login with invalid credentials

**Review by anatoly-devico — CHANGES_REQUESTED** [review]
Pay attention to all comments + try to make tests independent

File: cypress/e2e/new-transaction.spec.js [hunk 10]
```diff
@@ -1,97 +1,156 @@
 import {transaction} from "../selectors/new-transaction";
-import {another_user_info} from "../selectors/sign_in_page";
+
+const transactionAmount = "5"
+const noteText = "Sushi dinner"
 
 describe('New Transaction', () => {
```
Comment (anatoly-devico) [inline cypress/e2e/new-transaction.spec.js:6]: this spec doesn't work

**Review by anatoly-devico — APPROVED** [review]
Merge it and rearrange tests according to new specs structure

**Pull request merged**